calyx/src/backend/rtl/gen.rs

```diff
@@ -21,278 +22,252 @@ impl Backend for RtlBackend {
-//     //             .append(RcDoc::text(";")),
-//     //     ),
-//     //     Port::This { .. } => None,
-//     // }
-// }
+fn wire_string<'a>(
+    portdef: ast::Portdef,
+    structure: &'a structure::StructureGraph,
+    idx: NodeIndex,
+    name: &ast::Id,
+) -> Option<D<'a, ColorSpec>> {
+    // build comment of the destinations of each wire declaration
+    let dests: Vec<D<ColorSpec>> = structure
+        .connected_to(idx, portdef.name.to_string())
+        .map(|(node, edge)| {
+            D::text(node.get_name().to_string())
+                .append(" @ ")
+                .append(&edge.dest)
+        })
+        .collect();
+    if !dests.is_empty() {
+        let dest_comment =
+            D::text("// ").append(D::intersperse(dests, D::text(", ")));
+        let wire_name = format!("{}${}", &name, &portdef.name);
+        Some(
+            colors::keyword(D::text("wire"))
+                .append(D::space())
+                .append(bit_width(portdef.width))
+                .append(colors::ident(D::text(wire_name)))
+                .append(";")
+                .append(D::space())
+                .append(dest_comment),
+        )
+    } else {
+        None
+    }
+}
 
-// /**
-//  * Generates a string wirename for the provided Port object
-//  */
-// pub fn port_wire_id(p: &Port) -> RcDoc<'_> {
-//     match p {
-//         Port::Comp { component, port } => RcDoc::text(component)
-//             .append(RcDoc::text("_"))
-//             .append(RcDoc::text(port)),
-//         Port::This { port } => RcDoc::text(port),
-//     }
-// }
+pub fn bit_width<'a>(width: u64) -> D<'a, ColorSpec> {
+    use std::cmp::Ordering;
+    match width.cmp(&1) {
+        Ordering::Less => panic!("Invalid bit width!"),
+        Ordering::Equal => D::nil(),
+        Ordering::Greater => {
+            D::text(format!("[{}:0]", width - 1)).append(D::space())
+        }
+    }
+}
 
-// //==========================================
-// //        Subcomponent Instance Functions
-// //==========================================
-// // Intermediate data structures for string formatting
-// #[derive(Clone, Debug)]
-// pub struct RtlInst<'a> {
-//     pub comp_name: &'a String,
-//     pub id: &'a String,
-//     pub params: Vec<u64>,
-//     pub ports: HashMap<&'a String, String>, // Maps Port names to wires
-// }
+//==========================================
+//        Subcomponent Instance Functions
+//==========================================
+fn subcomponent_instances<'a>(comp: &component::Component) -> D<'a, ColorSpec> {
+    use crate::lang::structure::NodeData;
+    let subs = comp.structure.instances().filter_map(|(idx, data)| {
+        if let NodeData::Instance {
+            name,
+            structure,
+            signature,
+        } = data
+        {
+            let doc = subcomponent_sig(&name, &structure)
+                .append(D::space())
+                .append(parens(
+                    D::line()

```

Comment:
Instead of explicitly `append`ing everything, is it possible to defined `parens`, `line`, and `space` as methods on the the pretty printing object (I'm assuming `D<'a, ColorSpec>` is such an object).

Reply:
At the moment this is not possible because you can’t have impls on traits defined in a library. Soon, this will be possible in some capacity in rust and we can transition to that. The other option is implementing our own Doc type instead of using RcDoc but this seemed like more work that it was worth at the time.